pkg/controllers/pod_controller.go

```diff
@@ -0,0 +1,284 @@
+		log.Info("finding leader pod from list")
+		leaderPod = findLeaderPod(podList)
+
+		// If leader pod is not scheduled yet, do nothing and return.
+		if leaderPod == nil || leaderPod.Spec.NodeName == "" {
+			log.Info("leader pod not scheduled")
+			return ctrl.Result{}, nil
+		}
+
+		log.Info(fmt.Sprintf("follower pod %q is following leader %q", pod.Name, leaderPod.Name))
+	}
+
+	log.Info(fmt.Sprintf("getting topology value from topology key: %s", topologyKey))
+	topologyValue, err := r.topologyFromPod(ctx, leaderPod)
+	if err != nil {
+		log.Error(err, "getting topology value from the node the leader pod is running on")
+		return ctrl.Result{}, err
+	}
+
+	log.Info("assigning all job pods to same topology as leader")
+	if err := r.assignGatedPodsToTopology(ctx, podList, topologyKey, topologyValue); err != nil {
+		log.Error(err, "assigning job pods to topology")
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *PodReconciler) topologyFromPod(ctx context.Context, pod *corev1.Pod) (string, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	nodeName := pod.Spec.NodeName
+	ns := pod.Namespace
+	topologyKey := pod.Annotations[jobset.ExclusiveKey]
+
+	// Get node the leader pod is running on.
+	var node corev1.Node
+	if err := r.Get(ctx, types.NamespacedName{Name: nodeName, Namespace: ns}, &node); err != nil {
+		// We'll ignore not-found errors, since there is nothing we can do here.
+		// A node may not exist temporarily due to a maintenance event or other scenarios.
+		log.Error(err, fmt.Sprintf("getting node %s", nodeName))
+		return "", client.IgnoreNotFound(err)
+	}
+
+	// Get topology (e.g. node pool name) from node labels.
+	topology, exists := node.Labels[topologyKey]
+	if !exists {
+		return "", fmt.Errorf("node does not have topology label: %s", topology)
+	}
+	return topology, nil
+}
+
+func (r *PodReconciler) assignGatedPodsToTopology(ctx context.Context, podList *corev1.PodList, topologyKey, topologyValue string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// For every pod besides pod index 0, add a node selector for this topology and remove the scheduling gate.
+	// If we fail to update an individual pod, continue updating the remaining pods but return the error after
+	// to trigger reconciliation.
+	var finalErr error
+	for _, pod := range podList.Items {
+		if isLeaderPod(&pod) {
+			continue
+		}
+		// Skip this pod if scheduling gate was already removed. This means the pod reconciler
+		// has already processed it and assigned it to the same topology as the leader pod.
+		gateIndex := schedulingGateIndex(&pod)
+		if gateIndex == gateDoesNotExist {
+			continue
+		}
+
+		log.Info(fmt.Sprintf("adding node selector to pod: %s", pod.Name))
+		addNodeSelector(&pod, topologyKey, topologyValue)
+
+		log.Info(fmt.Sprintf("removing scheduling gate from pod: %s", pod.Name))
+		removeSchedulingGate(&pod, gateIndex)
+
+		log.Info(fmt.Sprintf("updating pod: %s", pod.Name))
+		if err := r.Update(ctx, &pod); err != nil {
```

Comment:
This is a blocking update and so a huge bottleneck, we need to do this in a go routine in parallel; if the update fails, we need to requeue and force another reconile. We need to rate limit re-queueing to avoid hammering the api-server if there is something really wrong.